src/crh/poseidon/mod.rs

```diff
@@ -31,191 +49,210 @@ pub trait PoseidonRoundParams<F: PrimeField>: Default + Clone {
         }
 
         // Finally the current_state becomes the output
-        current_state
-    }
-
-    pub fn hash_2(&self, xl: F, xr: F) -> F {
-        // Only 2 inputs to the permutation are set to the input of this hash
-        // function, one is set to the padding constant and rest are 0. Always keep
-        // the 1st input as 0
-        let input = vec![
-            F::from(ZERO_CONST),
-            xl,
-            xr,
-            F::from(PADDING_CONST),
-            F::from(ZERO_CONST),
-            F::from(ZERO_CONST),
-        ];
-
-        // Never take the first output
-        self.permute(&input)[1]
+        Ok(state)
     }
 
-    pub fn hash_4(&self, inputs: [F; 4]) -> F {
-        // Only 4 inputs to the permutation are set to the input of this hash
-        // function, one is set to the padding constant and one is set to 0. Always
-        // keep the 1st input as 0
-        let input = vec![
-            F::from(ZERO_CONST),
-            inputs[0],
-            inputs[1],
-            inputs[2],
-            inputs[3],
-            F::from(PADDING_CONST),
-        ];
-
-        // Never take the first output
-        self.permute(&input)[1]
+    fn apply_linear_layer(state: &Vec<F>, mds: &Vec<Vec<F>>) -> Vec<F> {
+        let mut new_state: Vec<F> = Vec::new();
+        for i in 0..state.len() {
+            let mut sc = F::zero();
+            for j in 0..state.len() {
+                let mij = mds[i][j];
+                sc += mij * state[j];
+            }
+            new_state.push(sc);
+        }
+        new_state
     }
 }
 
-pub struct PoseidonCRH<F: PrimeField, P: PoseidonRoundParams<F>> {
-    field: PhantomData<F>,
-    params: PhantomData<P>,
-}
+impl<F: PrimeField, P: Rounds> FixedLengthCRH for CRH<F, P> {
+    const INPUT_SIZE_BITS: usize = 32 * 8 * P::WIDTH;
+    type Output = F;
+    type Parameters = PoseidonParameters<F>;
 
-impl<F: PrimeField, P: PoseidonRoundParams<F>> PoseidonCRH<F, P> {
-    pub fn create_mds<R: Rng>(_rng: &mut R) -> Vec<Vec<F>> {
-        let mds_matrix = Vec::new();
-        mds_matrix
+    // Not sure what's the purpose of this function of we are going to pass parameters
+    fn setup<R: Rng>(rng: &mut R) -> Result<Self::Parameters, Error> {
+        Ok(Self::Parameters::generate(rng))
     }
 
-    pub fn create_round_consts<R: Rng>(_rng: &mut R) -> Vec<F> {
-        let round_consts = Vec::new();
-        round_consts
+    fn evaluate(parameters: &Self::Parameters, input: &[u8]) -> Result<Self::Output, Error> {
+        let eval_time = start_timer!(|| "PoseidonCRH::Eval");
+
+        if (input.len() * 8) > Self::INPUT_SIZE_BITS {
+            panic!(
+                "incorrect input length {:?} for width {:?}",
+                input.len() / 32,
+                P::WIDTH,
+            );
+        }
+
+        // Not giving expected results
+        // let elts: Vec<F> = input.to_field_elements().unwrap_or(Vec::new());
```

Comment:
What are the expected results here? `to_field_elements` packs a byte slice into the field elements as tightly as possible.

Reply:
Well, by "as tightly as possible", I mean that it follows the algorithm here: https://github.com/arkworks-rs/algebra/blob/e504bdafdbeb97fab86cc5726ee413b1d7236662/ff/src/to_field_vec.rs#L44

This algorithm rounds down the number of bytes in a field element. For example, if we're trying to pack 32 bytes into a Bn254::ScalarField, then it'll pack 31 bytes into the first field element, and 1 byte into a fresh field element.

Reply:
So is the solution to instead (if we are using BN254) input our elements as 31 byte elements to prevent against this? What other routes do you see to handle this @Pratyush ?

Reply:
The other way is to have a routine that handles the chunking automatically for you

Reply:
@Pratyush Can we have `ToBytes` impl return tightly packed bytes -- e.g. 31 bytes instead of 32 for Bn254? The natural flow is calling `ark_ff::to_bytes` to convert into bytes, and then `to_field_elements` to convert back to the field, but these two are not compatible

Reply:
In practice application protocols are going to have their own requirements for how the input/output is packed into field elements, if any packing is done at all. In Orchard for example we have the input and output as field elements, and the full range is needed.